_posts/2025-04-10-invoking-component-functions-in-wasmtime-cli.md

```diff
@@ -0,0 +1,215 @@
+
+## Default Entry Point vs. Exported Function
+
+If we were not using the component model and just creating binary executable, we could use a simple command like `wasmtime run foo.wasm` to execute our code.
+
+But in this case, we are making the point that Wasm has evolved beyond simple modules to embrace the Wasm Component Model. Developers need more granular control over the execution and composition of components. The benefits are code reuse and cross-language sharing. For example, Rust, Javascript, and Python logic can all interoperate when the explicit interfaces are defined using WIT. For example, [seamlessly interoperable compression tasks across native C++, Rust, and also in a Wasm runtime](https://medium.com/wasm/wasm-component-model-seamless-compression-c-rust-and-wasm-3b8d52ed8b31).
+
+## Invoking Exported Functions
+
+The addition of the `--invoke` feature (within the `wasmtime run` subcommand) allows users to specify an exported function to execute, complete with arguments, directly from the command line interface (CLI) terminal. Being able to invoke functions in the CLI is a huge step forward for not only executing Wasm via consoles or Bash scripting but will also play a big role in testing, debugging, and integrating Wasm components into workflows without having to embed them into a host application during development.
+
+## How Invoke Works: A Practical Example
+
+Originally, the `wasmtime run` command would take one positional argument (the name of the module) and just run that `.wasm` file:
+
+```bash
+wasmtime run foo.wasm
+```
+
+The `wasmtime run` command now accepts an optional `--invoke` argument, which can execute the name of an exported function that resides in the (`.wasm`) module:
+
+```bash
+wasmtime run --invoke "get-answer()" target/wasm32-wasip1/debug/wasm_answer.wasm
+```
+
+## Wasm Value Encoding (WAVE)
+
+Invoke leverages [WAVE](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasm-wave/README.md) as a standard way to encode function calls and/or results. WAVE is a human-oriented text encoding of Wasm Component Model values and is designed to be consistent with the [WIT IDL format](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md). Here are a few pointers for constructing your `wasmtime run --invoke` commands using WAVE.
+
+## Parentheses
+
+Parentheses after the exported function's name are mandatory. The presence of the parenthesis `()` signifies function invocation, as opposed to the function name just being referenced.
+
+## Double Quotes
+
+The exported function's name and mandatory exported function's parentheses must all be enclosed in one set of double quotes, i.e. `"get-answer()"`.
+
+The result from our correctly typed command above is as follows:
+
+```bash
+42
+```
+
+Note: If your function takes a string argument, ensure that you use escaped double quotes inside the parentheses. For example:
+
+```bash
+wasmtime run - invoke "initialize(\"hello\")" foo.wasm
```

Comment:
```suggestion
wasmtime run - invoke 'initialize("hello")' foo.wasm
```

Using single quotes on the outside avoids the need to escape inner double quotes.